Added Connection Resilliency to Driver

xcellsoft opened the pull request:
Connection resiliency has been added to the driver using the pattern that the EF Core SQL Server driver utilizes.

MySql Errors 2000-2027, 2047-2050 and 0 are automatically handled. If have others you would like to add or remove this can be done in the MySqlTransientExceptionDetector.cs. I tried to get all that would deal with transient connection/server issues.

**High level overview of the changes:**

**[New] EFCore.MySql\MySqlRetryingExecutionStrategy.cs**
This was copied straight over and name changed, the ShouldRetryOn was modified to deal with the MySqlException type.

**[Modified] EFCore.MySql\Extensions\MySqlServiceCollectionExtensions.cs**
Added the MySqlExecutionStrategyFactory to dependency injection.

**[Modified] EFCore.MySql\Infrastructure\MySqlDbContextOptionsBuilder.cs**

Added EnableRetryOnFailure option similar to SqlServer so we can do 3 varients when building the configuration options:

1. Specify number of retries, how long of delay between retries and additional errors to trap.

```
var addErrors = new Collection<int>();
addErrors.Add(1);
addErrors.Add(2);
options.UseMySql(tenantConnectionString.ConnectionString,
   o => o.EnableRetryOnFailure(2, TimeSpan.FromSeconds(5),addErrors));
```

2. Specify only number of retries.

```
options.UseMySql(tenantConnectionString.ConnectionString,
   o => o.EnableRetryOnFailure(2));
```

3. Default to system defined 6 retries.

```
options.UseMySql(tenantConnectionString.ConnectionString,
   o => o.EnableRetryOnFailure());
```

**[New] EFCore.MySql\Internal\MySqlExecutionStrategy.cs**
Straight copy with name change only

**[New ] EFCore.MySql\Internal\MySqlExecutionStrategyFactory.cs**
Straight copy with name change only

**[Modified] EFCore.MySql\Internal\MySqlOptions.cs**
Had to incorporate some special handling so that the connection operation called from this class could also have resilience as these direct calls to the database did not exist in the sql server one and are outside of the EF Core default executionstrategy. I see some other connect.Open() in this file which may need to be wrapped in the future.

**[New] EFCore.MySql\Internal\MySqlRetryNoDependendiciesExecutionStrategy.cs**
This is a special class to deal with the connection resiliency in the MySqlOptions.cs class without having access to ExecutionStrategyDependency.

**[New] EFCore.MySql\Internal\MySqlTransientExceptionDetector.cs**
New class copied from Sql Server implementation and configured to retry errors 2000-2027, 2047-2050 and 0.

**[Modified] EFCore.MySql\Storage\Internal\MySqlConnectionSettings.cs**
Added connection resiliency as this class falls outside of standard EF Core execution strategy.

xcellsoft commented:
Ignore Commit 2efed3ec6799d3b384fe32459df94f7300ee679d

It was used only to publish to nuget for temporary usage.

Kagamine commented on src/EFCore.MySql/EFCore.MySql.csproj:
```diff
@@ -1,16 +1,16 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <AssemblyName>Pomelo.EntityFrameworkCore.MySql</AssemblyName>
-    <PackageId>Pomelo.EntityFrameworkCore.MySql</PackageId>
+    <AssemblyName>XCell.EntityFrameworkCore.MySql</AssemblyName>
```
Revert this

xcellsoft replied:
Done.

Kagamine commented:
I would like this with an option in OptionBuilder to enable retry.

xcellsoft commented:
What are your thoughts?

If you don’t include the option it is basically backwards compatible with how it works currently.

I have put it through some testing, major reason for this is hosting with azure MySQL and dealing with any transient failures.

A lot of people will be interested with this, if I’m missing something let me know.

Thanks to you and your team for doing all the work you have already done.

xcellsoft commented:
Fyi, I mirrored the implementation that the ef core team did with sql server and the oracle sample that they did along with  what the postgres team did with this same implementation.

Kagamine commented:
@xcellsoft Okay, so could you give me the link, I wanna take a look at their implementation, thanks.

Kagamine commented:
Hi @caleblloyd, Could you take a look into this PR?

caleblloyd commented:
What is the default behavior? No retrying?

I think it would be a good feature to add, but keep in mind that SQL Server changes their internals a lot, so adding this means we have to keep up with their changes for more classes. That means we need a test to ensure it works with future upgrades also.

Can you add a test, similar to the test for Batch Size, that runs the functional tests with the retries enabled? Look in `.travis.yml` at the root of the repository.

xcellsoft commented:
I just added an executionstrategy which in itself is pretty benign an simple and follows the ef core standards. The extra part was to handle the non standard stuff with the same logic at the startup where you get the version by opening a connection. I do understand why it was done that way, without thinking connecting resiliency I would have done it that way too.

I think we are stuck with adapting to the changes that the ef core team does and we need to be in tune with them. This is similar to the other pull request I did where they made a breaking change which they should not have done, that change was a bad practice.

xcellsoft commented:
FYI, the Postgres team made the exact changes to their driver with the exception of what I did to handle the version connection. I think they let those connections to be handled within an executionstrategy handler.

Also, it looks like the ef core team is creating an Oracle driver and they implemented the same way. Look at the ef core github site.

xcellsoft commented:
Yes deafault behavior withou execution behavior is no retry. This is ef core behavior.

Kagamine commented:
I think at least we should make retry times as an option that user could modify it.

caleblloyd commented:
@xcellsoft I'm good with the change once the test is added.  Can you take a shot at?  In the `.travis.yml`?

xcellsoft commented:
I'll take a look at the .travis.yml file and see what i can do.

Responses to other questions:

@Kagamine , This is how you specify the number of retries, in my example i have it set to 2 tries with 5 seconds between each try and i'm not asking it to look at any additional errors. The does control this feature and it can only be enabled if the user invokes EnableRetryOnFailure 

```
options.UseMySql(tenantConnectionString.ConnectionString,
   o => o.EnableRetryOnFailure(2, TimeSpan.FromSeconds(5),null));
```

@Kagamine , The errors it traps by default are MySqlException Numbers: 0, 2000-2027, 2047 - 2050 which appear to be all the server related errors that could occur and that the hosted environment could be potentially cleaning up (managed MySql servers in cluster with auto-recovery). 

@Kagamine here are the links to this type of modification with the other drivers:

[Postgress Retry Execution Strategy](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/commit/201a45cbb63b072da40053ebcff2febee97b9b8c#diff-cde8a50a1ad617a0890ae8ee65e86116)

[EF Core Oracle Driver Sample - there is not a commit just for the Resiliency Part](https://github.com/aspnet/EntityFrameworkCore/tree/dev/samples/OracleProvider/src/OracleProvider)

xcellsoft commented:
@caleblloyd , 

I tried to get the test environment running locally so i could write the test, when I run the rebuild.ps1 i get a whole bunch of these errors below, it appears the ef migrations is creating open strings spanning multiple lines error. Let me know...thanks.

Error Example:
```
                    b.Property<string>("Name")
                        .ValueGeneratedOnAddOrUpdate()
                        .HasColumnType("VARCHAR(63) GENERATED ALWAYS AS (
						`Names` ->> \"$[0]\"
					) VIRTUAL");
```

```
Migrations\20171101132558_initial.Designer.cs(471,20): error CS1010: Newline in constant [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]
Migrations\AppDbModelSnapshot.cs(470,20): error CS1010: Newline in constant [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]
Migrations\20171101132558_initial.Designer.cs(472,8): error CS1002: ; expected [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]
Migrations\AppDbModelSnapshot.cs(471,8): error CS1002: ; expected [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]
Migrations\20171101132558_initial.Designer.cs(472,15): error CS1002: ; expected [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]
Migrations\AppDbModelSnapshot.cs(471,15): error CS1002: ; expected [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]
Migrations\20171101132558_initial.Designer.cs(472,15): error CS1010: Newline in constant [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]
Migrations\AppDbModelSnapshot.cs(471,15): error CS1010: Newline in constant [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]
Migrations\20171101132558_initial.Designer.cs(472,18): error CS1002: ; expected [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]
Migrations\AppDbModelSnapshot.cs(471,18): error CS1002: ; expected [D:\Git\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests\EFCore.MySql.FunctionalTests.csproj]

```
These are the steps I was following:

[Functional and Performance Tests](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/tree/master/test/EFCore.MySql.FunctionalTests)

> Next, you must rebuild migrations. Run the scripts/rebuild.sh script on Linux or the scripts/rebuild.ps1 script on Windows. Any time you make changes to database models, run the rebuild script.

xcellsoft commented:
Also, any advice on the best way to create a new AppDb data context as I'll need to do that to invoke the EnableRetryOnFailure() - I could store the connection string from the config in the startup.cs and then access it in the unit test. Thoughts?

Thanks.

```
options.UseMySql(tenantConnectionString.ConnectionString,
   o => o.EnableRetryOnFailure(2));
```

caleblloyd commented:
@xcellsoft my recommendation is that you don't create a DB Context, just make another test run.  Add EF_RETRY_ON_FAILURE environment variable here:

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/test/EFCore.MySql.FunctionalTests/AppConfig.cs

```
	    public static readonly int EfRetryOnFailure = !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("EF_RETRY_ON_FAILURE"))
		    ? Convert.ToInt32(Environment.GetEnvironmentVariable("EF_RETRY_ON_FAILURE")) : 0;
```

Then handle the environment variable in the Startup.cs here:

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/63fe3c3ba327cc8f4b19cc2809a481996443102b/test/EFCore.MySql.FunctionalTests/Startup.cs#L46-L60

Finally, add a test where you set the environment variable in `.travis.yml`:

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/.travis.yml

```
- echo "Test with EF_RETRY_ON_FAILURE=3 " && docker exec -it dotnet sh -c 'cd /dotnet/test/EFCore.MySql.FunctionalTests && export EF_RETRY_ON_FAILURE=3 && dotnet test -c Release'
```

Kagamine commented:
I've taken a look into PG's implementation. So for your contribution, LGTM. Thank you.

xcellsoft commented:
@caleblloyd ,

Any ideas on the errors i'm getting in my local test environment (see errors above).

xcellsoft commented:
@caleblloyd ,

So it looks like entity framwork is removing your @ symbol for multi-line for your values from GeneratedTypes.cs. I modified GeneratedTypes.cs to work without the @ and put everything on one line. This resolved my problem. Maybe there is another issue going on here with how the migrations are being built.

I changed this in GeneratedTypes.cs (yours was definitely more readable)

```
			modelBuilder.Entity<GeneratedContact>(entity =>
			{
				entity.HasIndex(m => m.Name);
				entity.HasIndex(m => m.Email);

				entity.Property(m => m.Name)
					.ValueGeneratedOnAddOrUpdate()
					.HasColumnType("VARCHAR(63) GENERATED ALWAYS AS (`Names` ->> \"$[0]\") VIRTUAL");

				entity.Property(m => m.Email)
					.ValueGeneratedOnAddOrUpdate()
					.HasColumnType("VARCHAR(63) GENERATED ALWAYS AS (`ContactInfo` ->> \"$.Email\") VIRTUAL");

				entity.Property(m => m.Address)
					.ValueGeneratedOnAddOrUpdate()
					.HasColumnType("VARCHAR(63) GENERATED ALWAYS AS (CONCAT_WS(', ',`ContactInfo` ->> \"$.Address\",`ContactInfo` ->> \"$.City\",`ContactInfo` ->> \"$.State\",`ContactInfo` ->> \"$.Zip\")) STORED");
			});
```

This is entity framework turned it into within the AppDbModelSnapshot:

```
                  b.Property<string>("Address")
                        .ValueGeneratedOnAddOrUpdate()
                        .HasColumnType("VARCHAR(63) GENERATED ALWAYS AS (
						CONCAT_WS(', ',
							`ContactInfo` ->> \"$.Address\",
                            `ContactInfo` ->> \"$.City\",
                            `ContactInfo` ->> \"$.State\",
                            `ContactInfo` ->> \"$.Zip\"
						)) STORED");

```

xcellsoft commented:
@caleblloyd ,

Unit test performed as you have outlined. Thanks.

caleblloyd commented:
I'm going to go ahead and merge and do some cleanup in a follow-up commit.

I believe that your problem with the `rebuild.ps1` script is related to your environment.  Mine runs just fine:

```
~\Projects\caleb\Pomelo.EntityFrameworkCore.MySql\test\EFCore.MySql.FunctionalTests [master ≡]> .\scripts\rebuild.ps1
Dropping database 'pomelo_test'.
Database 'pomelo_test' did not exist, no action was taken.
Done. To undo this action, use 'ef migrations remove'
Done.
```

Also, Appveyor uses this script and it runs just fine against `master` as well.

Thanks for the contribution!

caleblloyd commented:
This is in [2.0.1-preview-10075 on MyGet](https://www.myget.org/feed/pomelo/package/nuget/Pomelo.EntityFrameworkCore.MySql/2.0.1-preview-10075)

Kagamine commented on test/EFCore.MySql.FunctionalTests/Startup.cs:
```diff
@@ -47,15 +47,40 @@ public static void ConfigureEntityFramework(IServiceCollection services, DbConne
+                            mysqlOptions => 
+                            {
+                                mysqlOptions.MaxBatchSize(AppConfig.EfBatchSize);
+                                mysqlOptions.EnableRetryOnFailure(AppConfig.EfRetryOnFailure, TimeSpan.FromSeconds(5), null);
```
Need enable user input a retry timespan range, a fixed value is not good for retry.

caleblloyd replied:
Normally the user does configure this. It's just configured to 5 seconds for the functional tests which is fine.

xcellsoft replied:
Correct, the user can specify any timespan.

These options match exactly with sql server and PG

Kagamine replied:
No, my means is the time should not be fixed, user could provide 5~8seconds timespan, and ef will retry a random seconds in the 5~8 range.

xcellsoft replied:
I did some research on how the base EF Core libraries work with a registered ExecutionStrategy, the timeout setting actually does what you are outlining and then some. The maxRetryDelay is not actually the time between retries, it is the total time to retry the operation. The ExecutionStrategy class using this max amount, adds a little randomness and then spreads out the retries so there are more at the start and then exponentially it increases the time between retries. So for the most part we are good to go, except for the initial connection (see below).

What I have to work on is the custom routine I had to do (src/EFCore.MySql/Internal/MySqlRetryNoDependendiciesExecutionStrategy.cs)  to handle the connection that was outside of the ExecutionStrategy implementation in the src/EFCore.MySql/Storage/Internal/MySqlConnectionSettings.cs class. I used the maxRetryDelay as a time between retries as I thought that is what it was. I will correct.

I will post another pull request for this, I'm going to work on it now.

See the base EF Core ExecutionStrategy.cs for reference.
https://github.com/aspnet/EntityFrameworkCore/blob/09c9b263d648dac4432b9dfb280f84f651c33949/src/EFCore/Storage/ExecutionStrategy.cs

xcellsoft replied:
just submitted pull request for exponential backoff.

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/431
